### PR TITLE
Merge development into master

### PR DESCRIPTION
**Pull request by johnewart**
Side containers, remote build fixes, some build packs, logging improvements

### CODE REVIEW

File: cli/output.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package cli
```
[the remark this answers is not in the document]
Comment (johnewart) [reply cli/output.go]: Oh it gets used to setup the logging options (`SetupOutput` in `main.go`)

File: workspace/containers.go [hunk 2]
```diff
@@ -419,16 +602,16 @@ func (b BuildContainer) ExecToWriter(cmdString string, targetDir string, outputS
...
-	containerName := fmt.Sprintf("%s-%s", opts.Package.Name, containerImageName)
+	containerName := opts.containerName()
 
-	fmt.Printf("Creating container '%s'\n", containerName)
+	log.Infof("Creating container '%s'", containerName)
 
 	client := NewDockerClient()
 
-	PullImage(containerDef.Image)
+	if containerDef.Image == "" {
```
[the remark this answers is not in the document]
Comment (johnewart) [reply workspace/containers.go]: The pull happens after that

File: workspace/containers.go [hunk 3]
```diff
@@ -474,14 +657,20 @@ func NewContainer(opts BuildContainerOpts) (BuildContainer, error) {
...
-		internalPort := parts[1]
-		portKey := docker.Port(fmt.Sprintf("%s/tcp", internalPort))
-		var pb = make([]docker.PortBinding, 0)
-		pb = append(pb, docker.PortBinding{HostIP: "0.0.0.0", HostPort: externalPort})
-		bindings[portKey] = pb
+
+	if len(ports) > 0 {
+		log.Infof("Will map the following ports: ")
+
+		for _, portSpec := range containerDef.Ports {
```
[the remark this answers is not in the document]
Comment (johnewart) [reply workspace/containers.go]: Yes, we should sanity check them; though with our new "containers on networks" feature we probably can get rid of this anyway

File: workspace/containers.go [hunk 4]
```diff
@@ -526,116 +714,202 @@ func NewContainer(opts BuildContainerOpts) (BuildContainer, error) {
...
+			}
+		}
+	}
+
+	client := sc.DockerClient
+	if sc.NetworkId != "" {
+		log.Infof("Removing network...")
+		err := client.RemoveNetwork(sc.NetworkId)
+		if err != nil {
+			log.Warnf("Unable to remove network %s: %v", sc.NetworkId, err)
```
[the remark this answers is not in the document]
Comment (johnewart) [reply workspace/containers.go]: Sometime these fail but we need to try the others and it's not fatal, we just end up with networks we need to prune later.

File: cli/output.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (beholders-eye) [reply cli/output.go]: Only stderr output

**Pull request merged**